at_secondary/at_persistence_secondary_server/lib/src/compaction/compaction_service_impl.dart

```diff
@@ -0,0 +1,114 @@
+import 'package:at_persistence_secondary_server/at_persistence_secondary_server.dart';
+import 'package:at_persistence_secondary_server/src/compaction/compaction_service.dart';
+
+/// Class implementing the [CompactionService].
+/// On server start-up initializes the [_commitLogEntriesMap] with keys of [CommitEntry] from [CommitLogKeyStore].
+/// Further, on adding the entry to [CommitLogKeyStore] check's if the compaction threshold reaches.
+/// If yes, compacts the [CommitLogKeyStore].
+class CommitLogCompactionService implements CompactionService {
+  late CommitLogKeyStore _commitLogKeyStore;
+  final _commitLogEntriesMap = <String, CompactionSortedList>{};
+  int keysToCompactCount = 0;
+  List<AtCompactionLogObserver> observers = <AtCompactionLogObserver>[];
+
+  CommitLogCompactionService(CommitLogKeyStore commitLogKeyStore) {
+    _commitLogKeyStore = commitLogKeyStore;
+    commitLogKeyStore.toMap().then((map) => map.forEach((key, commitEntry) {
+          // If _commitLogEntriesMap contains the key, then more than one commitEntry exists.
+          // Increment the keysToCompactCount.
+          if (_commitLogEntriesMap.containsKey(key)) {
+            keysToCompactCount = keysToCompactCount + 1;
+          }
+          _commitLogEntriesMap.putIfAbsent(
+              commitEntry.atKey, () => CompactionSortedList());
+          _commitLogEntriesMap[commitEntry.atKey]!.add(key);
+        }));
+  }
+
+  @override
+  Future<void> informChange(CommitEntry commitEntry) async {
+    // If _commitLogEntriesMap contains the key, then more than one commitEntry exists.
+    // Increment the keysToCompactCount.
+    if (_commitLogEntriesMap.containsKey(commitEntry.atKey!)) {
+      keysToCompactCount = keysToCompactCount + 1;
+    }
+    _commitLogEntriesMap.putIfAbsent(
+        commitEntry.atKey!, () => CompactionSortedList());
+    _commitLogEntriesMap[commitEntry.atKey!]!.add(commitEntry.commitId!);
+    await compact();
+  }
+
+  @override
+  Future<void> compact() async {
+    if (_isCompactionRequired()) {
+      var keysBeforeCompaction = _commitLogKeyStore.getBox().length;
+      await Future.forEach(
+          _commitLogEntriesMap.keys, (key) => _compactExpiredKeys(key));
+      //Reset keysToCompactCount to 0 after deleting the expired keys.
+      keysToCompactCount = 0;
+      var keysAfterCompaction = _commitLogKeyStore.getBox().length;
+      // notify to observer when compaction is completed.
+      for (var observer in observers) {
+        observer.informCompletion(keysBeforeCompaction - keysAfterCompaction);
+      }
+    }
+  }
+
+  /// For the given atKey, gets the expired commitEntry keys and removes from keystore
+  void _compactExpiredKeys(var atKey) async {
+    var atKeyList = _commitLogEntriesMap[atKey];
+    var expiredKeys = atKeyList!.getKeysToCompact();
```

Comment:
This subList can be empty or null. I believe that's the reason you have a !.

Reply:
Update the code with null check. replaced ! with ?.